### PR TITLE
Issue #20: create install target for OpenIGTLinkIO

### PR DESCRIPTION
**Pull request by dkuegler**
create the INSTALL target, install target files, include files, Targets-file and Config file.
modify installed files to be machine independent.

### CODE REVIEW

**Comment by dkuegler** [comment]
Closed #20 in favor of this issue

**Comment by adamrankin** [comment]
So, a bit of a philosophical discussion: how can the installed files be machine independent? Do you mean, on machines that are using the same development environment, but with just different paths?

The reason I ask this, is because I've never seen the IS_INSTALL mechanic used. Is there no best practice CMake way of doing this?

File: OpenIGTLinkIOConfig.cmake.in [hunk 1]
```diff
@@ -1,23 +1,42 @@
...
-SET(OpenIGTLinkIO_INCLUDE_DIRS "@OpenIGTLinkIO_INCLUDE_DIRS@")
-SET(OpenIGTLinkIO_LIBRARY_DIRS "@OpenIGTLinkIO_LIBRARY_DIRS@")
 SET(OpenIGTLinkIO_LIBRARIES "@OpenIGTLinkIO_LIBRARIES@")
 
 IF(@IGTLIO_USE_GUI@)
   SET(IGTLIO_QT_VERSION "@IGTLIO_QT_VERSION@")
   IF(IGTLIO_QT_VERSION EQUAL "5")
-    SET(Qt5_DIR "@Qt5_DIR@")
+	IF(NOT DEFINED Qt5_DIR)
+	  SET(Qt5_DIR "@Qt5_DIR@")
```
Comment (adamrankin) [inline OpenIGTLinkIOConfig.cmake.in]: This opens up the possibility of multiple Qt5 versions used.

If you want add logic here, you could check if Qt5_FOUND and/or Qt5_DIR matches what this build of OpenIGTLinkIO was built with, and fire an error if mismatch.

File: Converter/CMakeLists.txt [hunk 2]
```diff
@@ -39,3 +40,11 @@ target_include_directories(${KIT} ${${KIT}_INCLUDE_DIRECTORIES})
 include( GenerateExportHeader )
 generate_export_header( ${KIT} EXPORT_FILE_NAME ${KIT}Export.h EXPORT_MACRO_NAME ${${KIT}_EXPORT_DIRECTIVE} )
 
+INSTALL(TARGETS ${KIT} EXPORT OpenIGTLinkIO 
+  RUNTIME DESTINATION "${OpenIGTLinkIO_BINARY_INSTALL}"
+  LIBRARY ARCHIVE DESTINATION "${OpenIGTLinkIO_LIBRARY_INSTALL}"
+  )
+
+INSTALL(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} DESTINATION "${OpenIGTLinkIO_INCLUDE_INSTALL}" PATTERN *.h)
```
Comment (adamrankin) [inline Converter/CMakeLists.txt]: Please enumerate the headers instead of using the PATTERN keyword.

**Comment by dkuegler** [comment]
Not best practice to my knowledge.
I can change it to use the "VTK-way", assuming VTK to use best practice.

File: OpenIGTLinkIOConfig.cmake.in (hunk 1 above)
Comment (dkuegler) [reply OpenIGTLinkIOConfig.cmake.in]: sounds reasonable

**Comment by dkuegler** [comment]
regarding machine dependency. I should really say build environment setup dependency.
The dependency on the compiler and target machine used cannot be removed. But the dependency on the specific folder setup can be removed.

**Comment by dkuegler** [comment]
I addressed the issues put forth here.

**Comment by adamrankin** [comment]
I have some changes, could I send them to you as a patch?

**Comment by dkuegler** [comment]
Sure, I’ll happily have a look and give my opinion.

Do you need my email?

**Comment by adamrankin** [comment]
[0001-Fixing-target-filename-capitalization.zip](https://github.com/IGSIO/OpenIGTLinkIO/files/751439/0001-Fixing-target-filename-capitalization.zip)

**Comment by adamrankin** [comment]
I think I created it properly? I've never created a git patch before, just SVN.

**Comment by adamrankin** [comment]
Hold off. The generated CMake Config.cmake for the install doesn't seem to be correct. Investigating.

**Comment by adamrankin** [comment]
This part is throwing me off:
get_filename_component(OpenIGTLinkIO_CMAKE_PACKAGE_DIR "${CMAKE_CURRENT_LIST_FILE}" DIRECTORY)
get_filename_component(OpenIGTLinkIO_INSTALL_PREFIX "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}" DIRECTORY)
get_filename_component(OpenIGTLinkIO_LIBRARY_DIR "${OpenIGTLinkIO_INSTALL_PREFIX}" DIRECTORY)
get_filename_component(OpenIGTLinkIO_INSTALL_PREFIX "${OpenIGTLinkIO_LIBRARY_DIR}" DIRECTORY)

What is OpenIGTLinkIO_LIBRARY_DIR supposed to be? The directory above the package dir? For people who use find_library(DIR SUFFIX /lib) ?

**Comment by dkuegler** [comment]
Well, typically the install will be into CMAKE_INSTALL_PREFIX and that will be “recovered into OpenIGTLinkIO_INSTALL_PREFIX” if you look at the typical targets file, the same pattern is typically found there as well, or in the typical vtk-install file.

Obviously expecting every installed file to be relative to CMAKE_INSTALL_PREFIX

So the CMAKE Config file will be ${CMAKE_INSTALL_PREFIX}/lib/igtlio/cmake/igtlio/OpenIGTLinkIOConfig.cmake

so OpenIGTLinkIO_CMAKE_PACKAGE_DIR should be ${CMAKE_INSTALL_PREFIX}/lib/igtlio/cmake/igtlio

so OpenIGTLinkIO_INSTALL_PREFIX should be ${CMAKE_INSTALL_PREFIX}/lib/igtlio/cmake

so OpenIGTLinkIO_CMAKE_PACKAGE_DIR should be ${CMAKE_INSTALL_PREFIX}/lib/igtlio

so OpenIGTLinkIO_LIBRARY_DIR should be ${CMAKE_INSTALL_PREFIX}/lib

so OpenIGTLinkIO_INSTALL_PREFIX should be ${CMAKE_INSTALL_PREFIX}

That is the thought process

**Comment by adamrankin** [comment]
Deleted.

**Comment by dkuegler** [comment]
So the patch actually builds and installs for me, so could you send me the exact error message?

**Comment by adamrankin** [comment]
Oh, but it breaks if it's not put into CMAKE_INSTALL_PREFIX

**Comment by adamrankin** [comment]
Shouldn't everything be relative to where the install directory is found?

I rewrote it to generate the following:

get_filename_component(OpenIGTLinkIO_CMAKE_PACKAGE_DIR "${CMAKE_CURRENT_LIST_FILE}" DIRECTORY)
get_filename_component(OpenIGTLinkIO_INSTALL_PREFIX "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}" DIRECTORY)

```
get_filename_component(OpenIGTLinkIO_CMAKE_PACKAGE_DIR "${CMAKE_CURRENT_LIST_FILE}" DIRECTORY)
get_filename_component(OpenIGTLinkIO_INSTALL_PREFIX "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}" DIRECTORY)

set(OpenIGTLinkIO_INCLUDE_DIR "${OpenIGTLinkIO_INSTALL_PREFIX}/include/igtlio")
set(OpenIGTLinkIO_BINARY_DIR "${OpenIGTLinkIO_INSTALL_PREFIX}/bin")
set(OpenIGTLinkIO_LIBRARY_DIR "${OpenIGTLinkIO_INSTALL_PREFIX}/@OpenIGTLinkIO_LIBRARY_INSTALL@")

set(OpenIGTLinkIO_INCLUDE_DIRS "${OpenIGTLinkIO_INCLUDE_DIR}" "${OpenIGTLinkIO_INCLUDE_DIR}/DeviceWidgets")
set(OpenIGTLinkIO_LIBRARY_DIRS "${OpenIGTLinkIO_BINARY_DIR} ${OpenIGTLinkIO_LIBRARY_DIR}")

include("${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}/lib/cmake/igtlio/OpenIGTLinkIOTargets.cmake")
```

(I didn't realize it was supposed to be /lib/igtlio/cmake/igtlio, qt has it directly as lib/cmake/...)

**Comment by dkuegler** [comment]
If what is not put into CMAKE_INSTALL_PREFIX … all installed files should “always” be placed into a subfolder of CMAKE_INSTALL_PREFIX (which can be “/”).

As I understand it, this is the intended way for it to work.

(the install macro actally places all files into folders relative to CMAKE_INSTALL_PREFIX)

**Comment by adamrankin** [comment]
Oh, but that's wrong too... it should be

```
get_filename_component(OpenIGTLinkIO_CMAKE_PACKAGE_DIR "${CMAKE_CURRENT_LIST_FILE}" DIRECTORY)
get_filename_component(OpenIGTLinkIO_INSTALL_PREFIX "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}" DIRECTORY)

set(OpenIGTLinkIO_INCLUDE_DIR "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}/include/igtlio")
set(OpenIGTLinkIO_BINARY_DIR "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}/bin")
set(OpenIGTLinkIO_LIBRARY_DIR "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}/@OpenIGTLinkIO_LIBRARY_INSTALL@")

set(OpenIGTLinkIO_INCLUDE_DIRS "${OpenIGTLinkIO_INCLUDE_DIR}" "${OpenIGTLinkIO_INCLUDE_DIR}/DeviceWidgets")
set(OpenIGTLinkIO_LIBRARY_DIRS "${OpenIGTLinkIO_BINARY_DIR} ${OpenIGTLinkIO_LIBRARY_DIR}")

include("${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}/lib/cmake/igtlio/OpenIGTLinkIOTargets.cmake")
```

**Comment by dkuegler** [comment]
Everything should be placed into CMAKE_INSTALL_PREFIX and relative to CMAKE_INSTALL_PREFIX, which is done, if relative paths are used for the install signature.

And then in the Config and Targets files should reference accordingly (they should be placed into a specific folder in there)

**Comment by dkuegler** [comment]
Maybe your confusing is, that the config file is not placed into the top folder and so the OpenIGTLinkIOConfig.cmake file is not placed into the top-level directory either.

**Comment by adamrankin** [comment]
Oh, so relocatable packages can't be placed anywhere?

**Comment by dkuegler** [comment]
They can be placed anywhere (as long as the file + paths names do not conflict with the filesystem specifications (esp. FAT has filename-length-limitations)

**Comment by dkuegler** [comment]
Lets change the first OpenIGTLinkIO_INSTALL_PREFIX to _TEMP_DIRECTORY

This is equivalent and might clear things up….


so OpenIGTLinkIO_CMAKE_PACKAGE_DIR should be ${CMAKE_INSTALL_PREFIX}/lib/igtlio/cmake/igtlio

so _TEMP_DIRECTORY should be ${CMAKE_INSTALL_PREFIX}/lib/igtlio/cmake

so OpenIGTLinkIO_CMAKE_PACKAGE_DIR should be ${CMAKE_INSTALL_PREFIX}/lib/igtlio

so OpenIGTLinkIO_LIBRARY_DIR should be ${CMAKE_INSTALL_PREFIX}/lib

so OpenIGTLinkIO_INSTALL_PREFIX should be ${CMAKE_INSTALL_PREFIX}


This part is throwing me off:
get_filename_component(OpenIGTLinkIO_CMAKE_PACKAGE_DIR "${CMAKE_CURRENT_LIST_FILE}" DIRECTORY)
get_filename_component(_TEMP_DIRECTORY "${OpenIGTLinkIO_CMAKE_PACKAGE_DIR}" DIRECTORY)
get_filename_component(OpenIGTLinkIO_LIBRARY_DIR "${_TEMP_DIRECTORY }" DIRECTORY)
get_filename_component(OpenIGTLinkIO_INSTALL_PREFIX "${OpenIGTLinkIO_LIBRARY_DIR}" DIRECTORY)

**Comment by adamrankin** [comment]
So, undoing my changes, and setting OpenIGTLinkIO_DIR (in PlusBuild configure) to 
`C:/d/PlusB-vs2015/Deps/OpenIGTLinkIO-bin/install`
produces the following error:
```
CMake Error at C:/d/PlusB-vs2015/Deps/OpenIGTLinkIO-bin/install/OpenIGTLinkIOConfig.cmake:25 (include):
  include could not find load file:

    C:/d/PlusB-vs2015/Deps/OpenIGTLinkIO-bin/install/OpenIGTLinkIOTargets.cmake
Call Stack (most recent call first):
  SuperBuild/External_OpenIGTLinkIO.cmake:3 (FIND_PACKAGE)
  CMakeLists.txt:388 (INCLUDE)
```

**Comment by dkuegler** [comment]
The main issue is, that the Config File should be placed into ${CMAKE_INSTALL_PREFIX}/lib/cmake/igtlio
But I put an extra igtlio as a path into the paths… complicating things. I’ll go ahead and fix that

**Comment by adamrankin** [comment]
> Well, typically the install will be into CMAKE_INSTALL_PREFIX and that will be “recovered into OpenIGTLinkIO_INSTALL_PREFIX” if you look at the typical targets file, the same pattern is typically found there as well, or in the typical vtk-install file.

Ah, I see. Unfortunately this is probably a reasonable assumption on Linux/Mac, but doesn't hold up on Windows.

I think this is the core issue. We cannot assume that the install is into CMAKE_INSTALL_PREFIX.

**Comment by dkuegler** [comment]
On Windows, you can „produce individual Folders for different packages“

So you would use “C:/Packages/OpenIGTLinkIO” as CMAKE_INSTALL_PREFIX, if you want clean folders with easy upgradeability or just “C:/Packages” if you do not mind the clutter. In the second case you can use CMAKE_PREFIX_PATH as “C:/Packages” to be able to autofind the OpenIGTLink_DIR variable.

**Comment by adamrankin** [comment]
This again assumes a single folder where all installs go. This is not a safe assumption. Is there no way to make the install completely self-referencing regardless of where it's been installed? 

It may mean that the user has to set the OpenIGTLinkIO_DIR to the install location, but that's a reasonable expectation.

**Comment by dkuegler** [comment]
There seems to be a missunderstanding.

This install is a completely self-referencing “installation”… this is the point of the install.

I just pushed a fix to my pull request, that tries fixing all your changes not related to the location of the Config file.

I will also install my install folder, so you can see how I do it… you should unpack it and

This is a install without SHARED LIBRARIES (that is why there is no bin folder).

**Comment by dkuegler** [comment]
Just as a reference: my PlusLib Project has the following Entry for

OpenIGTLinkIO_DIR

C:/Packages/OpenIGTLinkIO/lib/cmake/igtlio

I just zipped the folder C:/Packages/OpenIGTLinkIO for you.

Actually it is C:/Programs/OpenIGTLinkIO, but that makes NOT difference whatsoever. My Students have completely different folders including different drives and different folders. It is independent. (Well I started building OpenIGTLinkIO as a STATIC Library, so I would not need to distribute OpenIGTLinkIO any more it is just built into Plus, but anyway, that change is recent.


Typically I INSTALL my Packages into C:/Packages/<PackageName> and rename the Package to C:/Packages/<PackageName>-<Version> so I can archive it and use multiple Versions of the same Package at the same time (e.g. VTK-6.3.0 and VTK-7.1.0).
The renaming does not imply any change in any files, except I need to update using CMakeChache files to reference the folder including the Version.

**Comment by dkuegler** [comment]
(I didn't realize it was supposed to be /lib/igtlio/cmake/igtlio, qt has it directly as lib/cmake/...)

Which is correct and I changed that in my pull request.

**Comment by adamrankin** [comment]
Ignore.

**Comment by adamrankin** [comment]
Ooooooooooohhhhhhhhhh!!!! I'm referencing the wrong file.

Why does \<prefix\>/OpenIGTLinkIOConfig.cmake exist?

**Comment by dkuegler** [comment]
You created it and install does not „clean”, so this file is from an earalier build… look at the last modified date…

**Comment by adamrankin** [comment]
No, it is created by configure of OpenIGTLinkIO. If I delete it and try to build (without configuring), I get 
```
1>  CMake Error at cmake_install.cmake:59 (file):
1>    file INSTALL cannot find
1>    "C:/d/PlusB-vs2015/Deps/OpenIGTLinkIO-bin/install/OpenIGTLinkIOConfig.cmake".
```

**Comment by dkuegler** [comment]
That is a file, which is produced by configure, so you need to hit configure on OpenIGTLinkIO once so install/OpenIGTLinkIOConfig.cmake is created

**Comment by adamrankin** [comment]
Ah! It's a temporary file. Can we put it somewhere different? Like, \<build\>/CMakeFiles?

**Comment by adamrankin** [comment]
Aaahhh wonderful wonderful! This is great. Sorry for all the confusion, I've never developed a relocatable build. I will send my last patch with some minor fixes.

**Comment by dkuegler** [comment]
Sure, it is in „<build>/install“ as this file is for install, but I will happily put it into <build>/CMakeFiles/install …

**Comment by adamrankin** [comment]
Thanks so much for your patience.

[0001-Fixing-indentation-in-generated-config-code.zip](https://github.com/IGSIO/OpenIGTLinkIO/files/751643/0001-Fixing-indentation-in-generated-config-code.zip)

**Pull request merged**